json.md

```diff
@@ -0,0 +1,357 @@
+JSON string as the method argument.
+
+<!-- skip -->
+```dart
+Map<String, dynamic> user = JSON.decode(json);
+
+print('Howdy, ${user['name']}!');
+print('We sent the verification link to ${user['email']}.');
+```
+
+Unfortunately, `JSON.decode()` merely returns a `Map<String, dynamic>`, meaning
+that we do not know the types of the values until runtime. With this approach,
+we lose most of the statically typed language features: type safety,
+autocompletion and most importantly, compile-time exceptions. Our code can
+become instantly more error-prone.
+
+For example, whenever we access the `name` or `email` fields, we could quickly
+introduce a typo. A typo which our compiler does not know of since our entire
+JSON merely lives in a map structure.
+
+### Serializing JSON inside model classes
+
+We can combat the previously mentioned problems by introducing a plain model
+class, which we call `User`. Inside the `User` class, we have:
+
+* a `User.fromJson` constructor, for constructing a new `User` instance from a
+  map structure
+* a `toJson` method, which converts a `User` instance into a map.
+
+This way, the _calling code_ can now have type safety, autocompletion for the
+`name` and `email` fields and compile-time exceptions. If we make typos or treat
+the fields as `int`s instead of `String`s, our app will not even compile,
+instead of crashing on runtime.
+
+**user.dart**
+
+<!-- skip -->
+```dart
+class User {
+  final String name;
+  final String email;
+
+  User(this.name, this.email);
+
+  User.fromJson(Map<String, dynamic> json)
+      : name = json['name'],
+        email = json['email'];
+
+  Map<String, dynamic> toJson() {
+    return {
+      'name': name,
+      'email': email,
+    };
+  }
+}
+```
+
+Now the responsibility of the serialization logic is moved inside the model
+itself. With this new approach, we can deserialize a user quite easily.
+
+<!-- skip -->
+```dart
+Map userMap = JSON.decode(json);
+var user = new User.fromJson(userMap);
+
+print('Howdy, ${user.name}!');
+print('We sent the verification link to ${user.email}.');
+```
+
+To serialize a user, we just call the `toJson` method in the `User` class.
+
+<!-- skip -->
+```dart
+String json = JSON.encode(user.toJson());
+```
+
+This way, the calling code does not have to worry about JSON serialization at
+all. However, the model class still definitely has to. In a production app, we
+would want to be sure that the serialization works properly. In practice, the
+`User.fromJson` and `User.toJson` methods both need to have unit tests in place
+to verify correct behavior.
+
+Also, real-world scenarios are not usually that simple. It is unlikely that we
+can get by with such small JSON responses. Nested JSON objects are not that
+uncommon either.
+
+It would be nice if there were something that handled the JSON serialization for
+us. Luckily, there is!
+
+<a name="code-generation"></a>
+## Serializing JSON using json_serializable
```

Comment:
mention: code generation here (as we talk about it above as "code generation" and that's more what people are looking for. They don't know what json_serializable is , at this point)

For example: serializing JSON using code generation